pkg/ebpf/btf.go

```diff
@@ -22,139 +22,68 @@ import (
-// - Func: A function type, referencing a Func_Proto
-// - Int: 32-bit signed integer type
-// - Struct: A struct containing a single Int member
-// - Ptr: A pointer type
-// - Func_Proto (second occurrence): Function prototype with two parameters: an Int and a Ptr
-// - Func (second occurrence): Function type referencing the second Func_Proto
-//
-// These types are defined according to the BTF type encoding documented at
-// https://docs.kernel.org/bpf/btf.html#type-encoding
-//
-// The generated TypeSection will be used by eBPF programs to provide BTF
-// information about functions, e.g. enabling helper functions
-func typeSection() *pb.TypeSection {
-
-	types := []*pb.BtfType{}
-
-	// 1: Func_Proto
-	types = append(types, &pb.BtfType{
-		NameOff:    0x0,
-		Info:       0x0d000000,
-		SizeOrType: 0x0,
-		Extra: &pb.BtfType_Empty{
-			Empty: &pb.Empty{},
-		},
-	})
-
-	// 2: Func
-	types = append(types, &pb.BtfType{
-		NameOff:    0x1,
-		Info:       0x0c000000,
-		SizeOrType: 0x01,
-		Extra: &pb.BtfType_Empty{
-			Empty: &pb.Empty{},
-		},
-	})
-
-	// 3: Int
-	types = append(types, &pb.BtfType{
-		NameOff:    0x1,
-		Info:       0x01000000,
-		SizeOrType: 0x4,
-		Extra: &pb.BtfType_IntTypeData{
-			IntTypeData: &pb.IntTypeData{IntInfo: 0x01000020},
-		},
-	})
-
-	// 4: Struct
-	types = append(types, &pb.BtfType{
-		NameOff:    0x1,
-		Info:       0x04000001,
-		SizeOrType: 0x4,
-		Extra: &pb.BtfType_StructTypeData{
-			StructTypeData: &pb.StructTypeData{
-				NameOff:    0x1,
-				StructType: 0x3,
-				Offset:     0x0,
-			},
-		},
-	})
-
-	// 5: Ptr
-	types = append(types, &pb.BtfType{
-		NameOff:    0x0,
-		Info:       0x02000000,
-		SizeOrType: 0x4,
-		Extra: &pb.BtfType_Empty{
-			Empty: &pb.Empty{},
-		},
-	})
-
-	// 6: Func_Proto
-	types = append(types, &pb.BtfType{
-		NameOff:    0x0,
-		Info:       0x0d000002,
-		SizeOrType: 0x3,
-		Extra: &pb.BtfType_FuncProtoTypeData{
-			FuncProtoTypeData: &pb.FuncProtoTypeData{
-				Param: []*pb.BtfParam{
-					{NameOff: 0x1, ParamType: 0x3},
-					{NameOff: 0x1, ParamType: 0x5},
-				},
-			},
-		},
-	})
+// SetHeaderSection takes a BTF proto and assigns the header values, magic
+// represents if the system is big or little endian (0xeb9f, 0x9feb).
+func SetHeaderSection(btf *pb.Btf, magic int32, version int32, flags int32) {
+	btf.Header = &pb.Header{
+		Magic:   magic,
+		Version: version,
+		Flags:   flags,
+		HdrLen:  0x18,
+		TypeOff: 0x0,
+		TypeLen: int32(proto.Size(btf.TypeSection)),
+		StrOff:  int32(proto.Size(btf.TypeSection)),
+		StrLen:  int32(proto.Size(btf.StringSection)),
+	}
+}
 
-	// 7: Func
-	types = append(types, &pb.BtfType{
-		NameOff:    0x1,
-		Info:       0x0c000000,
-		SizeOrType: 0x6,
-		Extra: &pb.BtfType_Empty{
-			Empty: &pb.Empty{},
-		},
-	})
-	return &pb.TypeSection{BtfType: types}
+// SetTypeSection populates the BTF type section of the provided BTF proto with
+// the parameter types
+func SetTypeSection(btf *pb.Btf, types []*pb.BtfType) {
```

Comment:
do we need these setter methods if we are just adding a field to the proto? Can the caller just do it themselves?

Reply:
True, I'll change it .